polkadot/node/network/statement-distribution/src/v2/tests/requests.rs

```diff
@@ -1103,6 +1103,560 @@ fn peer_reported_for_providing_statements_with_wrong_validator_id() {
+			);
+		}
+
+		// Send statement from peer B.
+		let statement_b = state
+			.sign_statement(
+				v_b,
+				CompactStatement::Seconded(candidate_hash),
+				&SigningContext { parent_hash: relay_parent, session_index: 1 },
+			)
+			.as_unchecked()
+			.clone();
+		{
+			send_peer_message(
+				&mut overseer,
+				peer_b.clone(),
+				protocol_v2::StatementDistributionMessage::Statement(
+					relay_parent,
+					statement_b.clone(),
+				),
+			)
+			.await;
+
+			assert_matches!(
+				overseer.recv().await,
+				AllMessages::NetworkBridgeTx(NetworkBridgeTxMessage::ReportPeer(ReportPeerMessage::Single(p, r)))
+					if p == peer_b && r == BENEFIT_VALID_STATEMENT_FIRST.into() => { }
+			);
+
+			assert_matches!(
+				overseer.recv().await,
+				AllMessages::NetworkBridgeTx(NetworkBridgeTxMessage::SendValidationMessage(peers, _)) if peers == vec![peer_a]
+			);
+		}
+
+		// Send Backed notification.
+		{
+			overseer
+				.send(FromOrchestra::Communication {
+					msg: StatementDistributionMessage::Backed(candidate_hash),
+				})
+				.await;
+
+			assert_matches!(
+				overseer.recv().await,
+				AllMessages:: NetworkBridgeTx(
+					NetworkBridgeTxMessage::SendValidationMessage(
+						peers,
+						Versioned::V2(
+							protocol_v2::ValidationProtocol::StatementDistribution(
+								protocol_v2::StatementDistributionMessage::BackedCandidateManifest(manifest),
+							),
+						),
+					)
+				) => {
+					assert_eq!(peers, vec![peer_c]);
+					assert_eq!(manifest, BackedCandidateManifest {
+						relay_parent,
+						candidate_hash,
+						group_index: local_validator.group_index.unwrap(),
+						para_id: local_para,
+						parent_head_data_hash: pvd.parent_head.hash(),
+						statement_knowledge: StatementFilter {
+							seconded_in_group: bitvec::bitvec![u8, Lsb0; 1, 1, 1],
+							validated_in_group: bitvec::bitvec![u8, Lsb0; 0, 0, 0],
+						},
+					});
+				}
+			);
+
+			answer_expected_hypothetical_depth_request(&mut overseer, vec![], None, false).await;
+		}
+
+		let mask = StatementFilter {
+			seconded_in_group: bitvec::bitvec![u8, Lsb0; 0, 0, 1],
+			validated_in_group: bitvec::bitvec![u8, Lsb0; 0, 0, 0],
+		};
+
+		let relay_2 = Hash::repeat_byte(2);
+		let disabled_validators = vec![v_a];
+		let leaf_2 = state.make_dummy_leaf_with_disabled_validators(relay_2, disabled_validators);
+		activate_leaf(&mut overseer, &leaf_2, &state, false).await;
+		answer_expected_hypothetical_depth_request(&mut overseer, vec![], Some(relay_2), false)
+			.await;
+
+		// Incoming request to local node. Local node should not send the response as v_a is
+		// disabled and hence the quorum is not reached.
+		{
+			let response = state
+				.send_request(
+					peer_c,
+					request_v2::AttestedCandidateRequest { candidate_hash: candidate.hash(), mask },
+				)
+				.await
+				.await;
+
+			assert!(
+				response.is_none(),
+				"We should not send a response as the quorum is not reached yet"
+			);
```

Comment:
Nice. Does it make sense to test more edge cases, like local node sends out a request, finds out a validator is disabled, then receives a response? That test could also check if we stop sending out requests when quorum is no longer met (since as long as we don't get a response and the request is not outdated, we keep sending the request at [1 sec intervals](https://github.com/paritytech/polkadot-sdk/blob/mrcnski%2Fstatement-distribution-test-incoming-ack/polkadot/node/network/statement-distribution/src/v2/mod.rs#L134). But stopping such retries would just be an optimization).

Reply:
This was a good call!
I removed disabled mask from response management in [add disabled validators to unwanted mask](https://github.com/paritytech/polkadot-sdk/pull/1841/commits/40a8dce2016ac10b6f62d39d882227c32fc10aef) when addressing Rob's feedback. But actually, we still probably want to keep it, but not punish the peer for sending the statements in that case. Added a test that covers it. Thanks!

Reply:
@ordian Cool! Can you please point out where you added the test? I'm not able to see the git commit history.

Reply:
The test is called when_validator_disabled_after_sending_the_request:
https://github.com/paritytech/polkadot-sdk/pull/1841/files#diff-968d1fbf2f6253210d1620da40ad8591487e1bc54be0c123f5a7a59fc4c3938bR1228

Reply:
Test looks good overall, but I don't understand why we send out this Statement notification at the end. Why do we send a notification to a disabled peer?

https://github.com/paritytech/polkadot-sdk/blob/3a88190177a/polkadot/node/network/statement-distribution/src/v2/tests/requests.rs#L1364

Reply:
We send the confirmation about a different statement (`seconded_b`) to it. It's coming from `apply_post_confirmation` which calls `send_cluster_candidate_statements` (in this case peer_b is a cluster peer).

Reply:
Looks like we don't send it to `peer_b`, but the disabled peer. (Sending to `peer_b` wouldn't make sense either, it's the peer we received `seconded_b` from in the first place.)

```rs
assert_eq!(peers, vec![peer_disabled_later]);
```

Reply:
> Looks like we don't send it to peer_b, but the disabled peer.

Ah, right sorry. I looked at a different test.

Yes, we do send `seconded_b` statement to disabled peer, because it might not know about this statement. Disabling in statement-distribution works on the statement level, not peer level, so we still can send and receive statement from disabled validators, but not their statements. I think this is fine.

Reply:
Cool, resolving. :)